Replace Connection with MongoClient

mdbecker opened the pull request:
Old and busted; meet new hotness http://emptysquare.net/blog/pymongos-new-default-safe-writes/
`MongoClient` is new and shiny. As far as functionality goes, the API is backwards compatible with the old `Connection` class so this change shouldn't break anyone's code. The major change is that writes are safe by default. If you want to use unsafe writes, set `w=0`. If you use the `safe` kwarg in your code, you'll get warnings since it has been deprecated in favor of the `w` option. I've also replaced the existing `safe_insert` attribute with a new `write_concern_override` attribute. Unlike `safe_insert`, `write_concern_override` will impact all writes, not just calls to the save method. Usage of `safe_insert` will print a `DeprecationWarning`.

Commit:
connection: Replace Connection with MongoClient

Old and busted; meet new hotness http://emptysquare.net/blog/pymongos-new-default-safe-writes/
MongoClient is new and shiny. As far as functionality goes, the API is backwards compatible with the old Connection class so this change shouldn't break anyone's code. The major change is that writes are safe by default. If you want to use unsafe writes, set w=0. If you use the safe kwarg in your code, you'll get warnings since it has been deprecated in favor of the w option.

Commit:
setup.py: Updates for MongoClient

MongoClient requires pymongo 2.4 or greater.
In addition, let's increment the minor revision level to 0.2.0 to signify that this version of the package will break clients who require/expect unsafe writes but don't specify safe=False.

Commit:
docs.conf: Increment version from 0.1.8 to 0.2.0

klobmathews commented on setup.py:
```diff
@@ -18,7 +18,7 @@ def read(fname):
     license='BSD',
     packages=['scalymongo', 'scalymongo.manage'],
-    install_requires=['pymongo>=1.9'],
+    install_requires=['pymongo>=2.4'],
```
If you are going to upgrade pymongo, you should probably convert all safe=true --> w=1

scalymongo/document.py
tests/unit/test_document.py

mdbecker replied:
Good point.
All inserts default to safe so this logic is no longer necessary. I've gone ahead and removed the associated logic.
If you were previously setting this field as an arg instead of a kwarg, this change will break your code.
Additionally, the safe kwarg is deprecated, so you should switch to using the w flag instead.

Commit:
document: Replace safe_insert

The safe_insert attribute is deprecated in favor of the new write_concern_override attribute.
Unlike safe_insert, write_concern_override will impact all writes, not just calls to the save method.

I changed cls to mcs in the metaclass because mcs is the standard (see http://goo.gl/Cvo69)

Commit:
document: Unused import

Commit:
document: Don't use mutable default values

http://effbot.org/zone/default-values.htm

Commit:
document: Remove safe kwarg

The safe kwarg is deprecated, so you should switch to using the w flag instead.

mdbecker commented:
@allancaffee I added the deprecation warning. I also added new `write_concern` functionality in case this type of behavior is crucial to any users.
